### PR TITLE
Reimplement drag and drop

### PR DESCRIPTION
**Pull request by ramsays**
Initial reimplementation of drag and drop. Includes code to handle any null positions (from old data) and reset positions when they converge.

### CODE REVIEW

**Comment by vikaspotluri123** [comment]
Couple things:

1. Lint 😉 
2. It seems like category URLs are 404-ing on this branch. Tried on prod and I'm not running into any issues. i.e /app/semesters/2019F/courses/{id}. It works when using the UI, but once you refresh, you get 404'd
3. It seems like we will have to add an endpoint for batch update positions - my repro case is creating 30 grades, and rotating a couple of grades back and fourth. I think the current implementation is acceptable as a "hot fix" since it works for the hot path. In the merge, I did notice there are a couple of grades with the same position. Again, this can be handled later if needed

File: frontend/src/app/services/category.service.ts [hunk 1]
```diff
@@ -37,7 +37,12 @@ export class CategoryService {
 	}
 
 	get categories(): Category[] {
-		return this._categories[this._activeCourseID];
+		if (this._categories[this._activeCourseID]) {
```
Comment (vikaspotluri123) [inline frontend/src/app/services/category.service.ts]: I think this should probably be done at load time, and when a drag/drop event occurs

File: frontend/src/app/services/category.service.ts [hunk 2]
```diff
@@ -92,9 +97,9 @@ export class CategoryService {
 		}
 	}
 
-	async addCategory(name = ''): Promise<Category> {
+	async addCategory(data): Promise<Category> {
 		const createdCategory = transformRemoteCategory(
-			await this.api.createCategory({course: this._activeCourseID, name})
+			await this.api.createCategory(data)
```
Comment (vikaspotluri123) [inline frontend/src/app/services/category.service.ts]:
```suggestion
			await this.api.createCategory({course: this._activeCourseID, ...data});
```

**Comment by joshcosta** [comment]
I was able to create an unhandled collision of positions that causes the "compareCategory" function to break. This causes the course to become unusable. 

Repro: 
- Create course and add two categories
- Delete the first category (position 100)
- Try to add new category (Adds position 200, collision)

**Comment by joshcosta** [comment]
My commit made the initial position default to -Infinity. Surprisingly it seems to work, but I'll fix that.

File: frontend/src/app/gradetable/gradetable.component.ts [hunk 3]
```diff
@@ -120,9 +115,15 @@ export class GradetableComponent implements OnInit, OnDestroy {
 		this._courseChangeSubscription.unsubscribe();
 	}
 
-	async addGrade(): Promise<Category> {
-		console.log('gradetable.component::addGrade');
-		return this.categoryService.addCategory();
+	async addCategory(): Promise<Category> {
+		let maxPosition = 0;
+		if (this.categoryService.categories.length > 0) {
+			maxPosition = Math.max.apply(Math, this.categoryService.categories.map(function(cat) { return cat.position; }));
```
Comment (vikaspotluri123) [inline frontend/src/app/gradetable/gradetable.component.ts]:
```suggestion
			maxPosition = Math.max.apply(Math, this.categoryService.categories.map(({position}) => position));
```

**Comment by vikaspotluri123** [comment]
Got another collision bug:

1. Create 3 categories - A, B, C
2. Keep swapping A & B until it triggers a cleanup
3. Refresh page and look at `categories?course={id}` response - A and B both have a position of 200

**Comment by ramsays** [comment]
Just rewrote the reset handler function. It's a much more elegant design.

**Comment by vikaspotluri123** [comment]
Still getting duplicate positions on bulk-reset 😕

**Comment by ramsays** [comment]
Can't repro. Steps?

**Comment by vikaspotluri123** [comment]
Same as previous repro (https://github.com/gradebook/Aggie-Gradebook/pull/65#issuecomment-526861939)

**Pull request merged**